### PR TITLE
build-tools: Add several APIs to build-cli

### PR DESCRIPTION
**Pull request by tylerbutler**
Follow up to #11692.

Adds APIs for:

- Generating branch names.
- Creating release and dep bump branches.
- Checking npm for package updates.
- Checking if a particular version has been released.
- Getting all versions and their dates based on the git tag.
- Sorting versions based on date and semver version.

These APIs will be used heavily in the `release` and `release report` commands (future PRs).

### CODE REVIEW

File: build-tools/packages/build-cli/package.json [hunk 1]
```diff
@@ -44,16 +46,21 @@
   "dependencies": {
     "@fluid-tools/version-tools": "^0.4.3000",
     "@fluidframework/build-tools": "^0.4.3000",
+    "@fluidframework/test-version-utils": ">=2.0.0-internal.1.0.0 <2.0.0-internal.1.1.0",
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/package.json:49]:
```suggestion
```

File: build-tools/packages/build-cli/package.json [hunk 2]
```diff
@@ -51,9 +53,13 @@
...
     "@oclif/test": "^2",
     "chalk": "^2.4.2",
-    "inquirer": "^9.0.0",
+    "date-fns": "^2.29.1",
+    "extended-enum": "^1.0.1",
+    "inquirer": "^8.0.0",
+    "jssm": "^5.79.18",
     "npm-check-updates": "^16.0.0",
     "semver": "^7.3.7",
+    "semver-diff": "^4.0.0",
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/package.json:62]:
```suggestion
    "date-fns": "^2.29.1",
    "inquirer": "^8.0.0",
    "jssm": "^5.79.18",
    "npm-check-updates": "^16.0.0",
    "semver": "^7.3.7",
```

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+// eslint-disable-next-line max-params
+export async function npmCheckUpdates(
+    context: Context,
+    releaseGroup: ReleaseGroup | ReleasePackage,
+    depsToUpdate: ReleasePackage[] | RegExp[],
+    bumpType: "patch" | "minor" | "current",
+    // eslint-disable-next-line default-param-last
+    prerelease = false,
+    // eslint-disable-next-line default-param-last
+    writeChanges = false,
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/package.ts:62]: `writeChanges` is missing `@param` docs. Also the params list is out-of-order relative to the actual function paramter ordering.

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+ *
+ * @internal
+ */
+// eslint-disable-next-line max-params
+export async function npmCheckUpdates(
+    context: Context,
+    releaseGroup: ReleaseGroup | ReleasePackage,
+    depsToUpdate: ReleasePackage[] | RegExp[],
+    bumpType: "patch" | "minor" | "current",
+    // eslint-disable-next-line default-param-last
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/package.ts:59]: Not for this PR, but do we want this rule globally? I've always found this one odd, since optional parameters are really just syntactic sugar for the default value being undefined.

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+        releaseGroupOrPackage = context.repo.releaseGroups.get(releaseGroupOrPackage)!;
+    }
+
+    log?.verbose(`Checking for tag '${tagName}'`);
+    const rawTag = await context.gitRepo.getTags(tagName);
+    return rawTag.trim() === tagName;
+}
+
+/**
+ * Generates the correct tag name for the release of a given release group and version.
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/package.ts:290]: It might be worth explicitly calling out that this is a git tag. On first read it wasn't clear to me.,

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+/**
+ * Parses the version from a git tag.
+ *
+ * @param tag - The tag.
+ * @returns - The version string, or undefined if one could not be found.
+ *
+ * @internal
+ */
+export function getVersionFromTag(tag: string): string | undefined {
+    // TODO: Consider a stronger check here using a regex.
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/package.ts:353]: Did you intend to do this before merging, or later?

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+/**
+ * Sorts an array of {@link VersionDetails} by version or date. The array will be cloned then sorted in place.
+ *
+ * @param versions - The array of versions to sort.
+ * @param sortKey - The sort key.
+ * @returns A sorted array.
+ *
+ * @internal
+ */
+export async function sortVersions(versions: VersionDetails[], sortKey: "version" | "date") {
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/package.ts:436]: Missing return type

File: build-tools/packages/build-cli/src/lib/sets.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+ * @param setB - A set.
+ * @returns A new set containing the difference of Set A and Set B.
+ *
+ * @remarks
+ *
+ * Implementation from {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set}.
+ *
+ * @internal
+ */
+export function difference<T>(setA: Set<T>, setB: Set<T>) {
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/sets.ts:19]: Missing return type

Comment (Josmithr) [reply build-tools/packages/build-cli/src/lib/sets.ts:19]: Also, should we add this to one of our utils packages? Wouldn't give you access right now, but this seems useful more generally.

File: build-tools/packages/build-cli/src/releaseGroups.ts [hunk 9]
```diff
@@ -5,11 +5,29 @@
 
 import { isMonoRepoKind, MonoRepoKind } from "@fluidframework/build-tools";
 
+/**
+ * A type that represents independent packages (as opposed to those that are part of a release group).
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/releaseGroups.ts:9]: Is this the fully scoped package name?

File: build-tools/packages/build-cli/test/commands/generate/buildVersion.test.ts [hunk 10]
```diff
@@ -160,7 +160,7 @@ describe("generate:buildVersion", () => {
             "--build",
             "12345",
             "--fileVersion",
-            "1.2.4",
+            "1.2.5",
```
Comment (Josmithr) [inline build-tools/packages/build-cli/test/commands/generate/buildVersion.test.ts:163]: Just curious, why was this change necessary?

**Review by Josmithr — APPROVED** [review]

File: build-tools/packages/build-cli/src/base.ts [hunk 11]
```diff
@@ -58,7 +53,7 @@ export abstract class BaseCommand<T extends typeof BaseCommand.flags> extends Co
         return this.processedFlags as Partial<OutputFlags<typeof BaseCommand.flags>>;
     }
 
-    private _context: Context | undefined;
+    protected _context: Context | undefined;
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/base.ts:56]: This may not be needed any more. Changing back to private and if a future PR needs the change, I'll make it then.

File: build-tools/packages/build-cli/src/commands/bump/deps.ts [hunk 12]
```diff
@@ -207,7 +212,11 @@ export default class DepsCommand extends BaseCommand<typeof DepsCommand.flags> {
             const changedVersionMessage = changedVersionsString.join("\n");
             if (shouldCommit) {
                 const commitMessage = `Bump dependencies\n\n${changedVersionMessage}`;
-                const bumpBranch = `dep_${Date.now()}`;
+                const bumpBranch = generateBumpDepsBranchName(
+                    args.package_or_release_group,
+                    flags.bumpType!,
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/bump/deps.ts:217]: Added an explicit assert.

File: build-tools/packages/build-cli/src/lib/branches.ts [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+ * Generates an appropriate branch name from a release group, bump type, and version.
+ *
+ * @param releaseGroupOrPackage - The release group or independent package to generate a branch name for.
+ * @param bumpType - The bump type.
+ * @param version - The version to use for the generated branch name.
+ * @returns The generated branch name.
+ *
+ * @internal
+ */
+export function generateBumpBranchName(
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/lib/branches.ts:49]: Good idea.

File: build-tools/packages/build-cli/src/lib/branches.ts [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    fromInternalScheme,
+    VersionBumpType,
+    VersionBumpTypeExtended,
+} from "@fluid-tools/version-tools";
+import * as semver from "semver";
+import { isReleaseGroup, ReleaseGroup, ReleasePackage } from "../releaseGroups";
+import { getPackageShortName } from "./package";
+
+/**
+ * Creates an appropriate branch for a release group and bump type. Does not commit!
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/lib/branches.ts:19]: There are test cases that check/enforce the formats. I'll add more comments.

File: build-tools/packages/build-cli/src/lib/bump.ts [hunk 15]
```diff
@@ -22,47 +34,56 @@ export interface PackageWithRangeSpec {
  * value. If it is a bump type, the dependency range will be bumped according to that type.
  * @param prerelease - If true, will bump to the next pre-release version given the bump type.
  * @param onlyBumpPrerelease - If true, only dependencies on pre-release packages will be bumped.
+ * @param updateWithinSameReleaseGroup - If true, will update dependency ranges of deps within the same release group.
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/lib/bump.ts:37]: A case where this is useful is with our internal version scheme.

```ts
    // the lerna version command sets the dependency range of managed packages to a caret (^) dependency range. However,
    // for the internal version scheme, the range needs to be a >= < range.
```

In that case this function is called with that option set to true. I've added more comments, but do you think there's a better way to do this?

File: build-tools/packages/build-cli/src/lib/bump.ts [hunk 16]
```diff
@@ -73,3 +94,65 @@ export async function bumpPackageDependencies(
...
+/**
+ * Bumps a release group (or standalone package) by the bumpType.
+ *
+ * @param bumpType - The bump type.
+ * @param releaseGroupOrPackage - A release group repo or package to bump.
+ * @param scheme - The version scheme to use.
+ *
+ * @internal
+ */
+export async function bumpReleaseGroup(
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/lib/bump.ts:107]: Good idea.

File: build-tools/packages/build-cli/src/lib/package.ts (hunk 6 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/lib/package.ts:353]: Not in this PR, because I think it's more likely that I'll be too conservative with the regex, and with our tags right now the risk of the current code failing are low.

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 17]
```diff
@@ -0,0 +1,453 @@
...
+
+/**
+ * Returns the name of a package without the scope.
+ *
+ * @param pkgName - The full scoped package name.
+ * @returns The package name without the scope.
+ *
+ * @internal
+ */
+export function getPackageShortName(pkgName: string): string {
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/lib/package.ts:23]: Good point. The old code was very fluid-centric, and I made it generic enough we can replace it.

File: build-tools/packages/build-cli/test/commands/generate/buildVersion.test.ts (hunk 10 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/test/commands/generate/buildVersion.test.ts:163]: It technically wasn't. I was trying to resolve test failures but this wasn't the root cause. I'm just leaving it since the test is still valid.

Comment (tylerbutler) [reply build-tools/packages/build-cli/test/commands/generate/buildVersion.test.ts:163]: Actually reverted this after all.

File: build-tools/packages/build-cli/src/base.ts [hunk 18]
```diff
@@ -118,14 +112,37 @@ export abstract class BaseCommand<T extends typeof BaseCommand.flags> extends Co
...
+
+    public indent(indent = 2): string {
+        return " ".repeat(indent);
+    }
+
+    public errorLog(message: string | Error) {
+        this.log(chalk.red(`ERROR: ${message}`));
+    }
+
+    public warning(message: string | Error): string | Error {
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/base.ts]: left as-is because this overrides the oclif warning function.

File: build-tools/packages/build-cli/src/releaseGroups.ts (hunk 9 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/releaseGroups.ts:9]: Yes. Clarified.

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 19]
```diff
@@ -0,0 +1,440 @@
...
+            }
+
+            // Convert the range into the minimum version
+            const minVer = semver.minVersion(depVersion);
+            if (minVer === null) {
+                throw new Error(`semver.minVersion was null: ${depVersion} (${depName})`);
+            }
+
+            // If the min version has a pre-release section, then it needs to be released.
+            if (minVer.prerelease.length > 0) {
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/lib/package.ts:185]: Fixed.

File: build-tools/packages/build-cli/src/base.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (Josmithr) [reply build-tools/packages/build-cli/src/base.ts]: Oh, gotcha. Might be worth decorating with the `@override` tag to make that more immediately obvious, but that is a Josh nitpick 😁

**Review by Josmithr — APPROVED** [review]
